### PR TITLE
Common/CSW - Use status effect for `enableWeaponDisassembly`

### LINKED ISSUE #7842: ace_common_fnc_claim break CSW vanilla disassembly block
**Mods (complete and add to the following information):**
- **Arma 3:** `1.98` [stable]
- **CBA:** `3.15.1` [stable]
- **ACE3:** `3.13.3` [stable]

**Description:**
If vanilla disassembly turned off by CSW setting it can be turned back on by dragging pre-assembled weapon.

**Steps to reproduce:**
- Disable vanilla disassembly in CSW settings
- Place already assembled CSW static weapon
- Drag and drop it
- Action menu available now

**Expected behavior:**
Action menu should not be available

**Additional context:**
With a quick look at sources this behavior seems like happen because `ace_common_fnc_claim` do  
`_target enableWeaponDisassembly true;`
`ace_csw_assemblyMode`  is nil till you reassemble it. Here https://github.com/acemod/ACE3/blob/ebb1ef628f62b289371c31fda2b82b5af7fbdefe/addons/common/functions/fnc_claim.sqf#L33
So condition is true and `claim` reset `enableWeaponDisassembly`.

### PR DESCRIPTION
**Pull request by LinkIsGrim**
**When merged this pull request will:**
- Title.
- Close #7842

Ref #9304

### IMPORTANT

- [ ] If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/common/XEH_postInit.sqf [hunk 1]
```diff
@@ -28,6 +28,7 @@
 ["setHidden", true, ["ace_unconscious"]] call FUNC(statusEffect_addType);
 ["blockRadio", false, [QEGVAR(captives,Handcuffed), QEGVAR(captives,Surrendered), "ace_unconscious"]] call FUNC(statusEffect_addType);
 ["blockSpeaking", false, ["ace_unconscious"]] call FUNC(statusEffect_addType);
+["disableWeaponAssembly", false, ["ace_common", "ace_csw"]] call FUNC(statusEffect_addType);
```
Comment (PabstMirror) [inline addons/common/XEH_postInit.sqf:31]:
```suggestion
["disableWeaponAssembly", false, ["ace_common", "ace_csw_lockVehicle"]] call FUNC(statusEffect_addType);
```

File: addons/common/XEH_postInit.sqf [hunk 2]
```diff
@@ -154,10 +160,12 @@ if (isServer) then {
 [QGVAR(lockVehicle), {
     _this setVariable [QGVAR(lockStatus), locked _this];
     _this lock 2;
+    [_this, "disableWeaponAssembly", "ace_common", true] call FUNC(statusEffect_set);
```
Comment (PabstMirror) [inline addons/common/XEH_postInit.sqf:163]:
```suggestion
    if ([] isNotEqualTo getArray (configOf _this >> "assembleInfo" >> "dissasembleTo")) then {
        [_this, "disableWeaponAssembly", "ace_common_lockVehicle", true] call FUNC(statusEffect_set);
    };
```

File: addons/common/XEH_postInit.sqf [hunk 3]
```diff
@@ -154,10 +160,12 @@ if (isServer) then {
...
 }] call CBA_fnc_addEventHandler;
 
 [QGVAR(unlockVehicle), {
     _this lock (_this getVariable [QGVAR(lockStatus), locked _this]);
+    [_this, "disableWeaponAssembly", "ace_common", false] call FUNC(statusEffect_set);
```
Comment (PabstMirror) [inline addons/common/XEH_postInit.sqf:168]:
```suggestion
    if ([] isNotEqualTo getArray (configOf _target >> "assembleInfo" >> "dissasembleTo")) then {
        [_this, "disableWeaponAssembly", "ace_common_lockVehicle", false] call FUNC(statusEffect_set);
    };
```

File: addons/common/functions/fnc_claim.sqf [hunk 4]
```diff
@@ -29,19 +29,10 @@ if (!isNull _owner && {!isNull _unit} && {_unit != _owner}) then {
...
-    } else {
-        [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
-        if (_canBeDisassembled) then {
-            _target enableWeaponDisassembly true;
-        };
-    };
+if (_lockTarget && {!isNull _unit}) then {
+    [QGVAR(lockVehicle), _target, _target] call CBA_fnc_targetEvent;
+} else {
+    [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
```
Comment (PabstMirror) [inline addons/common/functions/fnc_claim.sqf:35]:
```suggestion
if (_lockTarget) then {
    if (!isNull _unit) then {
        [QGVAR(lockVehicle), _target, _target] call CBA_fnc_targetEvent;
    } else {
        [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
    };
```

I think it needs to be like this to keep old behaviour
or we'll call unlockVehicle everytime

File: addons/common/functions/fnc_claimSafeServer.sqf [hunk 5]
```diff
@@ -35,19 +35,10 @@ if (!isNull _owner && {!isNull _unit} && {_unit != _owner}) exitWith {
...
-    } else {
-        [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
-        if (_canBeDisassembled) then {
-            _target enableWeaponDisassembly true;
-        };
-    };
+if (_lockTarget && {!isNull _unit}) then {
+    [QGVAR(lockVehicle), _target, _target] call CBA_fnc_targetEvent;
+} else {
+    [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
```
Comment (johnb432) [inline addons/common/functions/fnc_claimSafeServer.sqf:41]: It isn't quite the same logic. Do you want a vehicle to be unlocked if `_unit` is `objNull`, despite `_lockTarget` being true?

```suggestion
if (_lockTarget) then {
    if (!isNull _unit) then {
        [QGVAR(lockVehicle), _target, _target] call CBA_fnc_targetEvent;
    } else {
        [QGVAR(unlockVehicle), _target, _target] call CBA_fnc_targetEvent;
    };
};
```

**Review by johnb432 — APPROVED** [review]
Apart from the comment, everything looks good to me. I don't have much knowledge about CSW, but the code is solid imo.

File: addons/csw/functions/fnc_assemble_deployWeapon.sqf [hunk 6]
```diff
@@ -46,8 +46,7 @@
             // Assembly mode: [0=disabled, 1=enabled, 2=enabled&unload, 3=default]
             _csw setVariable [QGVAR(assemblyMode), 2, true]; // Explicitly set advanced assembly mode + unload, and broadcast
             if (!GVAR(defaultAssemblyMode)) then {
-                TRACE_1("global disableVanillaAssembly event",_csw); // handles it being assembled when setting is disabled
-                [QGVAR(disableVanillaAssembly), [_csw]] call CBA_fnc_globalEvent;
+                [_csw, "disableWeaponAssembly", "ace_csw", true] call EFUNC(common,statusEffect_set);
```
Comment (LinkIsGrim) [inline addons/csw/functions/fnc_assemble_deployWeapon.sqf:49]:
```suggestion
                [_csw, "disableWeaponAssembly", QUOTE(ADDON), true] call EFUNC(common,statusEffect_set);
```

File: addons/csw/functions/fnc_staticWeaponInit.sqf [hunk 7]
```diff
@@ -49,8 +49,8 @@ if (_assemblyConfig) then {
         if (!alive _staticWeapon) exitWith { TRACE_1("dead/deleted",_staticWeapon); };
         private _assemblyMode = [false, true, true, GVAR(defaultAssemblyMode)] select (_staticWeapon getVariable [QGVAR(assemblyMode), 3]);
         TRACE_2("assemblyConfig present",_staticWeapon,_assemblyMode);
-        if (_assemblyMode) then { // Disable vanilla assembly if assemblyMode eanbled
-            [QGVAR(disableVanillaAssembly), [_staticWeapon]] call CBA_fnc_localEvent;
+        if (_assemblyMode) then { // Disable vanilla assembly if assemblyMode enabled
+            [_staticWeapon, "disableWeaponAssembly", "ace_csw", true] call EFUNC(common,statusEffect_set);
```
Comment (LinkIsGrim) [inline addons/csw/functions/fnc_staticWeaponInit.sqf:53]:
```suggestion
            [_staticWeapon, "disableWeaponAssembly", QUOTE(ADDON), true] call EFUNC(common,statusEffect_set);
```

File: addons/common/XEH_postInit.sqf [hunk 8]
```diff
@@ -161,10 +167,16 @@ if (isServer) then {
 [QGVAR(lockVehicle), {
     _this setVariable [QGVAR(lockStatus), locked _this];
     _this lock 2;
+    if ([] isNotEqualTo getArray (configOf _this >> "assembleInfo" >> "dissasembleTo")) then {
+        [_this, "disableWeaponAssembly", "ace_common_lockVehicle", true] call FUNC(statusEffect_set);
```
Comment (LinkIsGrim) [inline addons/common/XEH_postInit.sqf:171]:
```suggestion
        [_this, "disableWeaponAssembly", QGVAR(lockVehicle), true] call FUNC(statusEffect_set);
```

File: addons/common/XEH_postInit.sqf [hunk 9]
```diff
@@ -161,10 +167,16 @@ if (isServer) then {
...
+    };
 }] call CBA_fnc_addEventHandler;
 
 [QGVAR(unlockVehicle), {
     _this lock (_this getVariable [QGVAR(lockStatus), locked _this]);
+    if ([] isNotEqualTo getArray (configOf _target >> "assembleInfo" >> "dissasembleTo")) then {
+        [_this, "disableWeaponAssembly", "ace_common_lockVehicle", false] call FUNC(statusEffect_set);
```
Comment (LinkIsGrim) [inline addons/common/XEH_postInit.sqf:178]:
```suggestion
        [_this, "disableWeaponAssembly", QGVAR(lockVehicle), false] call FUNC(statusEffect_set);
```

**Pull request merged**